core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataDeleteIndexService.java

```diff
@@ -93,13 +116,20 @@ public ClusterState execute(final ClusterState currentState) {
                 MetaData.Builder metaDataBuilder = MetaData.builder(meta);
                 ClusterBlocks.Builder clusterBlocksBuilder = ClusterBlocks.builder().blocks(currentState.blocks());
 
+                final IndexGraveyard.Builder graveyardBuilder = metaDataBuilder.indexGraveyardBuilder();
                 for (final Index index : indices) {
                     String indexName = index.getName();
                     logger.debug("[{}] deleting index", index);
                     routingTableBuilder.remove(indexName);
                     clusterBlocksBuilder.removeIndexBlocks(indexName);
                     metaDataBuilder.remove(indexName);
+                    graveyardBuilder.addTombstone(index); // add the index tombstone to the cluster state
                 }
+                // purge tombstone entries if needed
+                final int numPurged = graveyardBuilder.purge(maxIndexTombstones);
+                logger.debug("{} tombstones purged from the cluster state.", numPurged);
```

Comment:
can we make it trace? and while at it - make sure you log the old size and the current size?